core/src/test/scala/core/NestedUnionsSpec.scala

```diff
@@ -0,0 +1,37 @@
+package core
+
+import org.scalatest.{FunSpec, Matchers}
+
+class NestedUnionsSpec extends FunSpec with Matchers with helpers.ApiJsonHelpers {
+
+    it("validates discriminator values are unique") {
+      def setup(userDiscriminatorValue: String, guestDiscriminatorValue: String) = {
+        val service = makeApiJson(
+          models = Map(
+            "user" -> makeModelWithField(),
+            "guest" -> makeModelWithField(),
+          ),
+          unions = Map(
+            "abstract_user" -> makeUnion(
+              types = Seq(makeUnionType("user", discriminatorValue = Some(userDiscriminatorValue)))
+            ),
+            "abstract_guest" -> makeUnion(
+              types = Seq(makeUnionType("guest", discriminatorValue = Some(guestDiscriminatorValue)))
+              ),
+            "party" -> makeUnion(
+              types = Seq(
+                makeUnionType("abstract_user"),
+                makeUnionType("abstract_guest"),
+              )
+            )
+          )
+        )
+        TestHelper.serviceValidator(service).errors()
+      }
+
+      setup("user", "guest") shouldBe Nil
+      setup("user", "user") shouldBe Seq(
+        "Union[party] discriminator value[user] appears more than once"
+      )
```

Comment:
This update adds a case where the discriminator of a union type itself is duplicated.

```suggestion
      def setup(
        userDiscriminatorValue: String,
        guestDiscriminatorValue: String,
        abtractUserDiscriminatorValue: String,
      ) = {
        val service = makeApiJson(
          models = Map(
            "user" -> makeModelWithField(),
            "guest" -> makeModelWithField(),
          ),
          unions = Map(
            "abstract_user" -> makeUnion(
              types = Seq(makeUnionType("user", discriminatorValue = Some(userDiscriminatorValue)))
            ),
            "abstract_guest" -> makeUnion(
              types = Seq(makeUnionType("guest", discriminatorValue = Some(guestDiscriminatorValue)))
            ),
            "party" -> makeUnion(
              types = Seq(
                makeUnionType("abstract_user", discriminatorValue = Some(abtractUserDiscriminatorValue)),
                makeUnionType("abstract_guest"),
              )
            )
          )
        )
        TestHelper.serviceValidator(service).errors()
      }

      setup("user", "guest", "abstract") shouldBe Nil
      setup("user", "user", "abstract") shouldBe Seq(
        "Union[party] discriminator value[user] appears more than once"
      )
      setup("user", "guest", "user") shouldBe Seq(
        "Union[party] discriminator value[user] appears more than once"
      )
```

Reply:
updated with:
```
  abstractUserDiscriminatorValue: Option[String],
```